test/test.py

```diff
@@ -0,0 +1,355 @@
+
+                gfxd_target(gfxd_f3dex2)
+                gfxd_endian(GfxdEndian.big, 4)
+
+                gfxd_execute()
+
+                self.assertEqual(expected, gfxd_buffer_to_string(outbuf))
+
+    def test_gfxd_output_buffer(self):
+        for sym, data, expected in self.data:
+            with self.subTest(sym):
+                gfxd_input_buffer(data)
+
+                outb = bytes(1000)
+                outbuf = gfxd_output_buffer(outb, len(outb))
+
+                gfxd_target(gfxd_f3dex2)
+                gfxd_endian(GfxdEndian.big, 4)
+
+                gfxd_execute()
+
+                self.assertEqual(expected, gfxd_buffer_to_string(outbuf))
+
+    def test_gfxd_output_fd(self):
+        for sym, data, expected in self.data:
+            with self.subTest(sym):
+                with tempfile.TemporaryFile() as output_file:
+                    gfxd_input_buffer(data)
+
+                    gfxd_output_fd(output_file)
+
+                    gfxd_target(gfxd_f3dex2)
+                    gfxd_endian(GfxdEndian.big, 4)
+
+                    gfxd_execute()
+
+                    output_file.seek(0)
+                    self.assertEqual(expected, output_file.read().decode())
+
+    def test_gfxd_output_callback(self):
+        for sym, data, expected in self.data:
+            with self.subTest(sym):
+                output = io.BytesIO()
+
+                def callback(buf, count):
+                    self.assertEqual(count, len(buf))
+                    output.write(buf)
+                    return count
+
+                gfxd_input_buffer(data)
+
+                gfxd_output_callback(callback)
+
+                gfxd_target(gfxd_f3dex2)
+                gfxd_endian(GfxdEndian.big, 4)
+
+                gfxd_execute()
+
+                self.assertEqual(expected, output.getvalue().decode())
+
+
+class TestArgumentCallback(unittest.TestCase):
+    def setUp(self):
+        self.data: list[Sym, memoryview] = []
+        for sym in TEST_DATA.syms:
+            data = bytes(TEST_DATA.data[sym.offset :][: sym.size])
+            self.data.append((sym, data))
+
+    def test_gfxd_vtx_callback(self):
+        for sym, data in self.data:
+            expected = {
+                "emptyDList": [],
+                "oneTriDList": [(0x42042069, 3)],
+                "setLights1DList": [],
+            }[sym.name]
+            with self.subTest(sym):
+                gfxd_input_buffer(data)
+
+                gfxd_target(gfxd_f3dex2)
+                gfxd_endian(GfxdEndian.big, 4)
+
+                verts = []
+
+                def callback(vtx, num):
+                    verts.append((vtx, num))
+                    return 0
+
+                try:
+                    gfxd_vtx_callback(callback)
+
+                    gfxd_execute()
+                finally:
+                    def callback(vtx, num):
+                        return 0
+                    gfxd_vtx_callback(callback)
```

Comment:
So this (try: ... finally: clear callback) is kind of bad I think, tests need to clear the state or it will mess with other states since libgfxd "keeps running"/is the same instance

Not sure what to do about it. "As long as it works" though is imo fine for tests

I only spotted this issue from adding this vtx callback test case, I should probably go over other callbacks and do the same

------

Also `gfxd_vtx_callback(None)` should eventually be made to work as you'd expect (afaict there currently is no way to unset a callback)